### PR TITLE
[next] Settings/Tenant

### PR DESCRIPTION
**Pull request by wyattjoh**
## What does this PR do?

Adds support for the tenant cache.

### CODE REVIEW

File: src/core/server/services/tenant/cache.ts [hunk 1]
```diff
@@ -1,31 +1,61 @@
...
+    this.mongo = mongo;
+
+    // Create a new client application ID. This prevents duplicated messages
+    // generated by this application from being handled as external messages
+    // as we should have already processed it.
+    this.clientApplicationID = uuid.v4();
 
     // Prepare the list of all tenant's maintained by this instance.
-    this.tenants = new DataLoader(ids => retrieveManyTenants(db, ids));
+    this.tenantsByID = new DataLoader(ids => {
```
Comment (alinhle) [inline src/core/server/services/tenant/cache.ts]: Change this to  a _property intializer_?

```ts
tenantsByID = new DataLoader(ids => {
  logger.debug({ ids: ids.length }, "now loading tenants");
  return retrieveManyTenants(this.mongo, ids);
});
```

File: src/core/server/services/tenant/cache.ts [hunk 2]
```diff
@@ -1,31 +1,61 @@
...
+      logger.debug({ ids: ids.length }, "now loading tenants");
+      return retrieveManyTenants(mongo, ids);
+    });
+    this.tenantsByDomain = new DataLoader(domains => {
```
Comment (alinhle) [inline src/core/server/services/tenant/cache.ts]: Same as above?

**Review by alinhle — APPROVED** [review]
Small comments, but rest looks good

File: src/core/server/services/tenant/cache.ts (hunk 1 above)
Comment (wyattjoh) [reply src/core/server/services/tenant/cache.ts]: Can you elaborate on this one?

**Pull request merged**